API KEY UPDATES: this commit allows you to set API KEYS as ENV variables

appleseed-iii opened the pull request:
Important:
1. you no longer need to recommit/redeploy when API Keys are rolled over
  you can simply update the API Keys in the your ENV in your deploy environment (fleek)
2. API Keys are no longer shared with the world via the github repo if
  external users want to run the app outside of localhost they need to bring their own keys
  (the app uses Ethers community key for localhost/dev environment)
3. keeping API Keys out of dev environment will allow you to whitelist domains
  (because you no longer need to worry about localhost)
4. updated README to explain the process to new devs (very minor)

appleseed-iii committed:
API KEY UPDATES: this commit allows you to set API KEYS as ENV variables

Important:
1. you no longer need to recommit/redeploy when API Keys are rolled over
  you can simply update the API Keys in the your ENV in your deploy environment (fleek)
2. API Keys are no longer shared with the world via the github repo if
  external users want to run the app outside of localhost they need to bring their own keys
  (the app uses Ethers community key for localhost/dev environment)
3. keeping API Keys out of dev environment will allow you to whitelist domains
  (because you no longer need to worry about localhost)
4. updated README to explain the process to new devs (very minor)

appleseed-iii commented:
@ohmygirth @lolchocotaco I don't know what your deploy process is for fleek, but I'm assuming you can easily set environment variables on there.

The environment variable names required for this build are listed in `.env.example`.

I plan on deploying a series of API improvements over the next week or so. This is just step 1, but wanted you to have a chance to think about it.

appleseed-iii commented on src/hooks/web3Context.tsx:
```diff
@@ -4,43 +4,30 @@ import { StaticJsonRpcProvider, JsonRpcProvider, Web3Provider } from "@etherspro
+} else {
+  // this is the ethers common API key, suitable for testing, not prod
+  ALCHEMY_ID_LIST = ["_gg7wSSi0KMBsdKnGVfHDueq6xMB9EkC"];
+}
```
What's the `NODE_ENV` on https://staging.olympusdao.finance?
If it's not `"production"` we should tweak this if statement so that staging uses the REACT_APP_ALCHEMY_IDS rather than the ethers common API key

Zayen-X commented on .env.example:
```diff
@@ -0,0 +1,9 @@
+# add your INFURA ID
+REACT_APP_INFURA_IDS=""
```
I love this shit and this is how we should handle deploy specific configs. :)

Zayen-X reviewed (changes requested):
I love the overall layout. 
I have on request for future sustianability. Let's make a helper method that handles anything/everything related to `process.env`. We should isolate environment access in that ONE function and use it consistently (error handling, defaults, etc.). 

It allows us to use the `.env` more flexibly throughout the app. :)

appleseed-iii committed:
Move process.env into helper Class

1. add src/helpers/Environment for helper Class
2. also remove dangling, unused INFURA_ID from constants

appleseed-iii committed:
also removing unused UserProvider

appleseed-iii commented on src/hooks/web3Context.tsx:
```diff
@@ -2,22 +2,23 @@ import React, { useState, ReactElement, useContext, useEffect, useMemo, useCallb
-if (process.env.REACT_APP_INFURA_IDS) {
-  INFURA_ID_LIST = process.env.REACT_APP_INFURA_IDS.split(" ");
+if (EnvHelper.env.REACT_APP_INFURA_IDS) {
+  INFURA_ID_LIST = EnvHelper.env.REACT_APP_INFURA_IDS.split(" ");
```
@lolchocotaco it just occurred to me that the `EnvHelper` I created is not exactly what you want.
Are you wanting me to encapsulate all of these if statements (like the above, referenced) into logic in the helper?

That way we don't have `if (EnvHelper.env...` & we can simply have `INFURA_ID_LIST = EnvHelper.INFURA_ID_LIST`

appleseed-iii committed:
reencapsulate all `process.env` access within `Environment.ts`

appleseed-iii commented on src/hooks/web3Context.tsx:
```diff
@@ -2,58 +2,29 @@ import React, { useState, ReactElement, useContext, useEffect, useMemo, useCallb
-  if (chainID === 1) return `https://eth-mainnet.alchemyapi.io/v2/${randomAlchemyID}`;
-  else if (chainID === 4) return `https://eth-rinkeby.alchemyapi.io/v2/aF5TH9E9RGZwaAUdUd90BNsrVkDDoeaO`; // unbanksy's
-}
+const ALCHEMY_ID_LIST = EnvHelper.getAlchemyAPIKeyList();
```
@lolchocotaco per your request, encapsulating `{process.env}` access within `EnvHelper`. Cleans up a lot of code.

appleseed-iii committed:
remove unused exports from constants.js

appleseed-iii committed:
Add Websocket Support

1. branched off of api-keys-env
2. adds websocket support to the same env implementation
3. load balances any Alchemy Keys and/or Websockets you provide
4. in Development environment Alchemy Keys revert to the Ethers
   community key
5. current websocket implementation has a little more latency than
   Alchemy. May need a little tuning of the ws server.

appleseed-iii commented:
https://github.com/OlympusDAO/olympus-frontend/commit/7127672a23df481165162a1d0ddb5fc15c8f9068 adds websocket support to this implementation. I kept it on a separate branch because I wasn't sure if you all would have thoughts on that.

Honestly the correct procedure is probably to merge it into this branch. And we can just leave the `REACT_APP_SELF_HOSTED_WEBSOCKETS` environment variable blank until it's ready to go.

My bad, now I've got prs on top of prs: https://github.com/OlympusDAO/olympus-frontend/pull/429 🤓

appleseed-iii committed:
add jsdoc for getSelfHostedWebSockets

appleseed-iii committed:
Merge pull request #429 from OlympusDAO/api-keys-env-w-websocket-support

Add Websocket Support

appleseed-iii committed:
replace string.split params from string whitespace to regex whitespace

appleseed-iii committed:
merge and cleanup

appleseed-iii commented:
https://github.com/OlympusDAO/olympus-frontend/pull/415/commits/1747a81ce0dd55394839e7a645f552d672339008 and https://github.com/OlympusDAO/olympus-frontend/pull/415/commits/d457bb552ac3989c85561063176ca3da6e0158af replace hard code string whitespace with regex

appleseed-iii committed:
Merge branch 'develop' into api-keys-env

appleseed-iii committed:
Merge dangling develop commit

appleseed-iii committed:
update api-keys-env for `alpha` / `staging` ready

1. remove lots of dangling, unused constants from constants.ts
2. update env.example to note the wss necessity for production
3. update types in Environment.ts
4. check for empty API keys & log `console.error` if ALL are empty
5. move API_URI building into Environment.ts where the API_KEYs are built

Zayen-X reviewed (approved):
LGTM, pending LUSD bond release.

Zayen-X commented on README.md:
```diff
@@ -12,10 +12,15 @@ Required:
+# set up your environment variables
+$ cp .env.example .env
+
+# fill in your own values in .env, then =>
```
Can you add something about going to https://www.alchemy.com/ for the Alchemy values?

appleseed-iii replied:
added commentary in https://github.com/OlympusDAO/olympus-frontend/pull/415/commits/536f552bb114551f82ac711dd82a77355986246f

Zayen-X commented on .env.example:
```diff
@@ -0,0 +1,26 @@
+####### API KEYS #######
```
Can you clearly mark which ones are `OPTIONAL`. For example we definitely need `REACT_APP_ALCHEMY_IDS` but INFURA_IDS and REACT_APP_SELF_HOSTED_WEBSOCKETS doesn't have to exist.

appleseed-iii replied:
literally none of them are required in development environment https://github.com/OlympusDAO/olympus-frontend/pull/415/files#r714443937

appleseed-iii replied:
added commentary in. https://github.com/OlympusDAO/olympus-frontend/pull/415/commits/536f552bb114551f82ac711dd82a77355986246f

appleseed-iii committed:
Merge branch 'develop' into api-keys-env

appleseed-iii commented on src/helpers/Environment.ts:
```diff
@@ -0,0 +1,101 @@
+   * in development environment will return the `ethers` community api key so that devs don't need to add elements to their .env
+   * @returns Array of Alchemy API URIs or empty set
+   */
+  static getAlchemyAPIKeyList() {
+    let ALCHEMY_ID_LIST: string[];
+
+    // split the provided API keys on whitespace
+    if (
+      EnvHelper.env.NODE_ENV !== "development" &&
+      EnvHelper.env.REACT_APP_ALCHEMY_IDS &&
+      EnvHelper.isNotEmpty(EnvHelper.env.REACT_APP_ALCHEMY_IDS)
+    ) {
+      ALCHEMY_ID_LIST = EnvHelper.env.REACT_APP_ALCHEMY_IDS.split(EnvHelper.whitespaceRegex);
+    } else {
+      // this is the ethers common API key, suitable for testing, not prod
+      ALCHEMY_ID_LIST = ["_gg7wSSi0KMBsdKnGVfHDueq6xMB9EkC"];
+    }
```
@lolchocotaco line 39 -> in development environment ALCHEMY_ID_LIST will automatically have the `ethers` common API key added. So even if a dev has a COMPLETELY BLANK .env file... the app will still run properly.

appleseed-iii committed:
update notes in readme & .env.example for clarity

Zayen-X committed:
remove debug log
